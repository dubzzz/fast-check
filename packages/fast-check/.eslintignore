bundle/
node_modules/
lib/
lib-*/
perf/
jest*.js
jest*.cjs