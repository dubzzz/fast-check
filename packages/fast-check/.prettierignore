bundle/
coverage/
docs/
documentation/Arbitraries.md
lib/
lib-*/