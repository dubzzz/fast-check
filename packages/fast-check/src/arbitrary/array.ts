import type { Arbitrary } from '../check/arbitrary/definition/Arbitrary';
import { ArrayArbitrary } from './_internals/ArrayArbitrary';
import type { SizeForArbitrary } from './_internals/helpers/MaxLengthFromMinLength';
import {
  MaxLengthUpperBound,
  maxGeneratedLengthFromSizeForArbitrary,
} from './_internals/helpers/MaxLengthFromMinLength';
import type { DepthIdentifier } from './_internals/helpers/DepthContext';

/**
 * Constraints to be applied on {@link array}
 * @remarks Since 2.4.0
 * @public
 */
export interface ArrayConstraints {
  /**
   * Lower bound of the generated array size
   * @defaultValue 0
   * @remarks Since 2.4.0
   */
  minLength?: number;
  /**
   * Upper bound of the generated array size
   * @defaultValue 0x7fffffff â€” _defaulting seen as "max non specified" when `defaultSizeToMaxWhenMaxSpecified=true`_
   * @remarks Since 2.4.0
   */
  maxLength?: number;
  /**
   * Define how large the generated values should be (at max)
   *
   * When used in conjonction with `maxLength`, `size` will be used to define
   * the upper bound of the generated array size while `maxLength` will be used
   * to define and document the general maximal length allowed for this case.
   *
   * @remarks Since 2.22.0
   */
  size?: SizeForArbitrary;
  /**
   * When receiving a depth identifier, the arbitrary will impact the depth
   * attached to it to avoid going too deep if it already generated lots of items.
   *
   * In other words, if the number of generated values within the collection is large
   * then the generated items will tend to be less deep to avoid creating structures a lot
   * larger than expected.
   *
   * For the moment, the depth is not taken into account to compute the number of items to
   * define for a precise generate call of the array. Just applied onto eligible items.
   *
   * @remarks Since 2.25.0
   */
  depthIdentifier?: DepthIdentifier | string;
}

/**
 * Extra but internal constraints that can be passed to array.
 * This extra set is made of constraints mostly targets experimental and internal features not yet mature to be exposed.
 * @internal
 */
export interface ArrayConstraintsInternal<T> extends ArrayConstraints {
  /**
   * Extra user-definable and hardcoded values.
   * Each entry in the array could be used to build the final generated value outputed by the arbitrary of array on generate.
   * Each entry must have at least one element of type T into it.
   * Each T must be a value acceptable for the arbitrary passed to the array.
   */
  experimentalCustomSlices?: T[][];
}

/**
 * For arrays of values coming from `arb`
 *
 * @param arb - Arbitrary used to generate the values inside the array
 * @param constraints - Constraints to apply when building instances (since 2.4.0)
 *
 * @remarks Since 0.0.1
 * @public
 */
function array<T>(arb: Arbitrary<T>, constraints: ArrayConstraints = {}): Arbitrary<T[]> {
  const size = constraints.size;
  const minLength = constraints.minLength || 0;
  const maxLengthOrUnset = constraints.maxLength;
  const depthIdentifier = constraints.depthIdentifier;
  const maxLength = maxLengthOrUnset !== undefined ? maxLengthOrUnset : MaxLengthUpperBound;
  const specifiedMaxLength = maxLengthOrUnset !== undefined;
  const maxGeneratedLength = maxGeneratedLengthFromSizeForArbitrary(size, minLength, maxLength, specifiedMaxLength);
  const customSlices = (constraints as ArrayConstraintsInternal<T>).experimentalCustomSlices || [];
  return new ArrayArbitrary<T>(arb, minLength, maxGeneratedLength, maxLength, depthIdentifier, undefined, customSlices);
}
export { array };
