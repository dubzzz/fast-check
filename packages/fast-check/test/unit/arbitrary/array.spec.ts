import * as fc from 'fast-check';
import { array } from '../../../src/arbitrary/array';

import { FakeIntegerArbitrary, fakeArbitrary } from './__test-helpers__/ArbitraryHelpers';

import * as ArrayArbitraryMock from '../../../src/arbitrary/_internals/ArrayArbitrary';
import {
  assertProduceCorrectValues,
  assertProduceSameValueGivenSameSeed,
  assertProduceValuesShrinkableWithoutContext,
  assertShrinkProducesSameValueWithoutInitialContext,
  assertShrinkProducesStrictlySmallerValue,
} from './__test-helpers__/ArbitraryAssertions';
import { isStrictlySmallerArray } from './__test-helpers__/ArrayHelpers';
import { Value } from '../../../src/check/arbitrary/definition/Value';
import { buildShrinkTree, renderTree } from './__test-helpers__/ShrinkTree';
import { sizeRelatedGlobalConfigArb } from './__test-helpers__/SizeHelpers';
import { withConfiguredGlobal } from './__test-helpers__/GlobalSettingsHelpers';

function beforeEachHook() {
  jest.resetModules();
  jest.restoreAllMocks();
  fc.configureGlobal({ beforeEach: beforeEachHook });
}
beforeEach(beforeEachHook);

describe('array', () => {
  it('should instantiate ArrayArbitrary(arb, 0, ?, 0x7fffffff, n.a) for array(arb)', () => {
    fc.assert(
      fc.property(sizeRelatedGlobalConfigArb, (config) => {
        // Arrange
        const { instance: childInstance } = fakeArbitrary<unknown>();
        const { instance } = fakeArbitrary<unknown[]>();
        const ArrayArbitrary = jest.spyOn(ArrayArbitraryMock, 'ArrayArbitrary');
        ArrayArbitrary.mockImplementation(() => instance as ArrayArbitraryMock.ArrayArbitrary<unknown>);

        // Act
        const arb = withConfiguredGlobal(config, () => array(childInstance));

        // Assert
        expect(ArrayArbitrary).toHaveBeenCalledWith(
          childInstance,
          0,
          expect.any(Number),
          0x7fffffff,
          undefined,
          undefined,
          [],
        );
        const receivedGeneratedMaxLength = ArrayArbitrary.mock.calls[0][2]; // Expecting the real value would check an implementation detail
        expect(receivedGeneratedMaxLength).toBeGreaterThan(0);
        expect(receivedGeneratedMaxLength).toBeLessThanOrEqual(2 ** 31 - 1);
        expect(Number.isInteger(receivedGeneratedMaxLength)).toBe(true);
        expect(arb).toBe(instance);
      }),
    );
  });

  it('should instantiate ArrayArbitrary(arb, 0, ?, maxLength, n.a) for array(arb, {maxLength})', () => {
    fc.assert(
      fc.property(sizeRelatedGlobalConfigArb, fc.nat({ max: 2 ** 31 - 1 }), (config, maxLength) => {
        // Arrange
        const { instance: childInstance } = fakeArbitrary<unknown>();
        const { instance } = fakeArbitrary<unknown[]>();
        const ArrayArbitrary = jest.spyOn(ArrayArbitraryMock, 'ArrayArbitrary');
        ArrayArbitrary.mockImplementation(() => instance as ArrayArbitraryMock.ArrayArbitrary<unknown>);

        // Act
        const arb = withConfiguredGlobal(config, () => array(childInstance, { maxLength }));

        // Assert
        expect(ArrayArbitrary).toHaveBeenCalledWith(
          childInstance,
          0,
          expect.any(Number),
          maxLength,
          undefined,
          undefined,
          [],
        );
        const receivedGeneratedMaxLength = ArrayArbitrary.mock.calls[0][2]; // Expecting the real value would check an implementation detail
        expect(receivedGeneratedMaxLength).toBeGreaterThanOrEqual(0);
        expect(receivedGeneratedMaxLength).toBeLessThanOrEqual(maxLength);
        expect(Number.isInteger(receivedGeneratedMaxLength)).toBe(true);
        if (config.defaultSizeToMaxWhenMaxSpecified) {
          expect(ArrayArbitrary).toHaveBeenCalledWith(childInstance, 0, maxLength, maxLength, undefined, undefined, []);
        }
        expect(arb).toBe(instance);
      }),
    );
  });

  it('should instantiate ArrayArbitrary(arb, minLength, ?, 0x7fffffff, n.a) for array(arb, {minLength})', () => {
    fc.assert(
      fc.property(sizeRelatedGlobalConfigArb, fc.nat({ max: 2 ** 31 - 1 }), (config, minLength) => {
        // Arrange
        const { instance: childInstance } = fakeArbitrary<unknown>();
        const { instance } = fakeArbitrary<unknown[]>();
        const ArrayArbitrary = jest.spyOn(ArrayArbitraryMock, 'ArrayArbitrary');
        ArrayArbitrary.mockImplementation(() => instance as ArrayArbitraryMock.ArrayArbitrary<unknown>);

        // Act
        const arb = withConfiguredGlobal(config, () => array(childInstance, { minLength }));

        // Assert
        expect(ArrayArbitrary).toHaveBeenCalledWith(
          childInstance,
          minLength,
          expect.any(Number),
          0x7fffffff,
          undefined,
          undefined,
          [],
        );
        const receivedGeneratedMaxLength = ArrayArbitrary.mock.calls[0][2]; // Expecting the real value would check an implementation detail
        if (minLength !== 2 ** 31 - 1) {
          expect(receivedGeneratedMaxLength).toBeGreaterThan(minLength);
          expect(receivedGeneratedMaxLength).toBeLessThanOrEqual(2 ** 31 - 1);
          expect(Number.isInteger(receivedGeneratedMaxLength)).toBe(true);
        } else {
          expect(receivedGeneratedMaxLength).toEqual(minLength);
        }
        expect(arb).toBe(instance);
      }),
    );
  });

  it('should instantiate ArrayArbitrary(arb, minLength, ?, maxLength, n.a) for array(arb, {minLength,maxLength})', () => {
    fc.assert(
      fc.property(
        sizeRelatedGlobalConfigArb,
        fc.nat({ max: 2 ** 31 - 1 }),
        fc.nat({ max: 2 ** 31 - 1 }),
        (config, aLength, bLength) => {
          // Arrange
          const [minLength, maxLength] = aLength < bLength ? [aLength, bLength] : [bLength, aLength];
          const { instance: childInstance } = fakeArbitrary<unknown>();
          const { instance } = fakeArbitrary<unknown[]>();
          const ArrayArbitrary = jest.spyOn(ArrayArbitraryMock, 'ArrayArbitrary');
          ArrayArbitrary.mockImplementation(() => instance as ArrayArbitraryMock.ArrayArbitrary<unknown>);

          // Act
          const arb = withConfiguredGlobal(config, () => array(childInstance, { minLength, maxLength }));

          // Assert
          expect(ArrayArbitrary).toHaveBeenCalledWith(
            childInstance,
            minLength,
            expect.any(Number),
            maxLength,
            undefined,
            undefined,
            [],
          );
          const receivedGeneratedMaxLength = ArrayArbitrary.mock.calls[0][2]; // Expecting the real value would check an implementation detail
          expect(receivedGeneratedMaxLength).toBeGreaterThanOrEqual(minLength);
          expect(receivedGeneratedMaxLength).toBeLessThanOrEqual(maxLength);
          expect(Number.isInteger(receivedGeneratedMaxLength)).toBe(true);
          if (config.defaultSizeToMaxWhenMaxSpecified) {
            expect(ArrayArbitrary).toHaveBeenCalledWith(
              childInstance,
              minLength,
              maxLength,
              maxLength,
              undefined,
              undefined,
              [],
            );
          }
          expect(arb).toBe(instance);
        },
      ),
    );
  });

  it('should instantiate ArrayArbitrary(arb, minLength, ?, maxLength, identifier) for array(arb, {minLength,maxLength, depthIdentifier})', () => {
    fc.assert(
      fc.property(
        sizeRelatedGlobalConfigArb,
        fc.nat({ max: 2 ** 31 - 1 }),
        fc.nat({ max: 2 ** 31 - 1 }),
        fc.string(),
        (config, aLength, bLength, depthIdentifier) => {
          // Arrange
          const [minLength, maxLength] = aLength < bLength ? [aLength, bLength] : [bLength, aLength];
          const { instance: childInstance } = fakeArbitrary<unknown>();
          const { instance } = fakeArbitrary<unknown[]>();
          const ArrayArbitrary = jest.spyOn(ArrayArbitraryMock, 'ArrayArbitrary');
          ArrayArbitrary.mockImplementation(() => instance as ArrayArbitraryMock.ArrayArbitrary<unknown>);

          // Act
          const arb = withConfiguredGlobal(config, () =>
            array(childInstance, { minLength, maxLength, depthIdentifier }),
          );

          // Assert
          expect(ArrayArbitrary).toHaveBeenCalledWith(
            childInstance,
            minLength,
            expect.any(Number),
            maxLength,
            depthIdentifier,
            undefined,
            [],
          );
          const receivedGeneratedMaxLength = ArrayArbitrary.mock.calls[0][2]; // Expecting the real value would check an implementation detail
          expect(receivedGeneratedMaxLength).toBeGreaterThanOrEqual(minLength);
          expect(receivedGeneratedMaxLength).toBeLessThanOrEqual(maxLength);
          expect(Number.isInteger(receivedGeneratedMaxLength)).toBe(true);
          if (config.defaultSizeToMaxWhenMaxSpecified) {
            expect(ArrayArbitrary).toHaveBeenCalledWith(
              childInstance,
              minLength,
              maxLength,
              maxLength,
              depthIdentifier,
              undefined,
              [],
            );
          }
          expect(arb).toBe(instance);
        },
      ),
    );
  });

  it('should throw when minimum length is greater than maximum one', () => {
    fc.assert(
      fc.property(
        sizeRelatedGlobalConfigArb,
        fc.nat({ max: 2 ** 31 - 1 }),
        fc.nat({ max: 2 ** 31 - 1 }),
        (config, aLength, bLength) => {
          // Arrange
          fc.pre(aLength !== bLength);
          const [minLength, maxLength] = aLength < bLength ? [bLength, aLength] : [aLength, bLength];
          const { instance: childInstance } = fakeArbitrary<unknown>();

          // Act / Assert
          expect(() =>
            withConfiguredGlobal(config, () => array(childInstance, { minLength, maxLength })),
          ).toThrowError();
        },
      ),
    );
  });
});

describe('array (integration)', () => {
  type Extra = { minLength?: number; maxLength?: number };
  const extraParameters: fc.Arbitrary<Extra> = fc
    .tuple(fc.nat({ max: 5 }), fc.nat({ max: 30 }), fc.boolean(), fc.boolean())
    .map(([min, gap, withMin, withMax]) => ({
      minLength: withMin ? min : undefined,
      maxLength: withMax ? min + gap : undefined,
    }));

  const isCorrect = (value: number[], extra: Extra) => {
    if (extra.minLength !== undefined) {
      expect(value.length).toBeGreaterThanOrEqual(extra.minLength);
    }
    if (extra.maxLength !== undefined) {
      expect(value.length).toBeLessThanOrEqual(extra.maxLength);
    }
    for (const v of value) {
      expect(typeof v).toBe('number');
    }
  };

  const isStrictlySmaller = isStrictlySmallerArray;

  const arrayBuilder = (extra: Extra) => array(new FakeIntegerArbitrary(), extra);

  it('should produce the same values given the same seed', () => {
    assertProduceSameValueGivenSameSeed(arrayBuilder, { extraParameters });
  });

  it('should only produce correct values', () => {
    assertProduceCorrectValues(arrayBuilder, isCorrect, { extraParameters });
  });

  it('should produce values seen as shrinkable without any context', () => {
    assertProduceValuesShrinkableWithoutContext(arrayBuilder, { extraParameters });
  });

  it('should be able to shrink to the same values without initial context', () => {
    assertShrinkProducesSameValueWithoutInitialContext(arrayBuilder, { extraParameters });
  });

  it('should preserve strictly smaller ordering in shrink', () => {
    assertShrinkProducesStrictlySmallerValue(arrayBuilder, isStrictlySmaller, { extraParameters });
  });

  it.each`
    rawValue                 | minLength
    ${[2, 4, 8, 16, 32, 64]} | ${undefined}
    ${[2, 4, 8]}             | ${undefined}
    ${[2, 4, 8]}             | ${2}
    ${[2, 4, 8]}             | ${3}
  `('should be able to shrink $rawValue given constraints minLength:$minLength', ({ rawValue, minLength }) => {
    // Arrange
    const constraints = { minLength };
    const arb = array(new FakeIntegerArbitrary(0, 1000), constraints);
    const value = new Value(rawValue, undefined);

    // Act
    const renderedTree = renderTree(buildShrinkTree(arb, value, { numItems: 100 })).join('\n');

    // Assert
    expect(arb.canShrinkWithoutContext(rawValue)).toBe(true);
    expect(renderedTree).toMatchSnapshot();
  });
});
