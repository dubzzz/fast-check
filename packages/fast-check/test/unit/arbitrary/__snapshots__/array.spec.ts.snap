// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`array (integration) should be able to shrink [2, 4, 8, 16, 32, 64] given constraints minLength:undefined 1`] = `
"[2,4,8,16,32,64]
├> []
├> [16,32,64]
|  ├> [32,64]
|  |  ├> [64]
|  |  |  ├> []
|  |  |  ├> [62]
|  |  |  |  ├> []
|  |  |  |  ├> [59]
|  |  |  |  |  ├> []
|  |  |  |  |  ├> [55]
|  |  |  |  |  |  ├> []
|  |  |  |  |  |  ├> [50]
|  |  |  |  |  |  |  ├> []
|  |  |  |  |  |  |  ├> [44]
|  |  |  |  |  |  |  |  ├> []
|  |  |  |  |  |  |  |  ├> [37]
|  |  |  |  |  |  |  |  |  ├> []
|  |  |  |  |  |  |  |  |  ├> [29]
|  |  |  |  |  |  |  |  |  |  ├> []
|  |  |  |  |  |  |  |  |  |  ├> [20]
|  |  |  |  |  |  |  |  |  |  |  ├> []
|  |  |  |  |  |  |  |  |  |  |  ├> [10]
|  |  |  |  |  |  |  |  |  |  |  |  ├> []
|  |  |  |  |  |  |  |  |  |  |  |  └> [0]
|  |  |  |  |  |  |  |  |  |  |  |     └> []
|  |  |  |  |  |  |  |  |  |  |  └> [11]
|  |  |  |  |  |  |  |  |  |  |     ├> []
|  |  |  |  |  |  |  |  |  |  |     ├> [0]
|  |  |  |  |  |  |  |  |  |  |     |  └> []
|  |  |  |  |  |  |  |  |  |  |     └> [1]
|  |  |  |  |  |  |  |  |  |  |        └> []
|  |  |  |  |  |  |  |  |  |  └> [21]
|  |  |  |  |  |  |  |  |  |     ├> []
|  |  |  |  |  |  |  |  |  |     ├> [11]
|  |  |  |  |  |  |  |  |  |     |  ├> []
|  |  |  |  |  |  |  |  |  |     |  ├> [0]
|  |  |  |  |  |  |  |  |  |     |  |  └> []
|  |  |  |  |  |  |  |  |  |     |  └> [1]
|  |  |  |  |  |  |  |  |  |     |     └> []
|  |  |  |  |  |  |  |  |  |     └> [12]
|  |  |  |  |  |  |  |  |  |        ├> []
|  |  |  |  |  |  |  |  |  |        ├> [1]
|  |  |  |  |  |  |  |  |  |        |  └> []
|  |  |  |  |  |  |  |  |  |        └> [2]
|  |  |  |  |  |  |  |  |  |           └> []
|  |  |  |  |  |  |  |  |  └> [30]
|  |  |  |  |  |  |  |  |     ├> []
|  |  |  |  |  |  |  |  |     ├> [21]
|  |  |  |  |  |  |  |  |     |  ├> []
|  |  |  |  |  |  |  |  |     |  ├> [11]
|  |  |  |  |  |  |  |  |     |  |  ├> []
|  |  |  |  |  |  |  |  |     |  |  ├> [0]
|  |  |  |  |  |  |  |  |     |  |  |  └> []
|  |  |  |  |  |  |  |  |     |  |  └> [1]
|  |  |  |  |  |  |  |  |     |  |     └> []
|  |  |  |  |  |  |  |  |     |  └> [12]
|  |  |  |  |  |  |  |  |     |     ├> []
|  |  |  |  |  |  |  |  |     |     ├> [1]
|  |  |  |  |  |  |  |  |     |     |  └> []
|  |  |  |  |  |  |  |  |     |     └> [2]
|  |  |  |  |  |  |  |  |     |        └> []
|  |  |  |  |  |  |  |  |     └> [22]
|  |  |  |  |  |  |  |  |        ├> []
|  |  |  |  |  |  |  |  |        ├> [12]
|  |  |  |  |  |  |  |  |        |  ├> []
|  |  |  |  |  |  |  |  |        |  ├> [1]
|  |  |  |  |  |  |  |  |        |  |  └> []
|  |  |  |  |  |  |  |  |        |  └> [2]
|  |  |  |  |  |  |  |  |        |     └> []
|  |  |  |  |  |  |  |  |        └> [13]
|  |  |  |  |  |  |  |  |           ├> []
|  |  |  |  |  |  |  |  |           ├> [2]
|  |  |  |  |  |  |  |  |           |  └> []
|  |  |  |  |  |  |  |  |           └> [3]
|  |  |  |  |  |  |  |  |              └> []
|  |  |  |  |  |  |  |  └> [38]
|  |  |  |  |  |  |  |     ├> []
|  |  |  |  |  |  |  |     ├> [30]
|  |  |  |  |  |  |  |     |  ├> []
|  |  |  |  |  |  |  |     |  ├> [21]
|  |  |  |  |  |  |  |     |  |  ├> []
|  |  |  |  |  |  |  |     |  |  ├> [11]
|  |  |  |  |  |  |  |     |  |  |  ├> []
|  |  |  |  |  |  |  |     |  |  |  ├> [0]
|  |  |  |  |  |  |  |     |  |  |  |  └> []
|  |  |  |  |  |  |  |     |  |  |  └> [1]
|  |  |  |  |  |  |  |     |  |  |     └> []
|  |  |  |  |  |  |  |     |  |  └> [12]
|  |  |  |  |  |  |  |     |  |     ├> []
|  |  |  |  |  |  |  |     |  |     ├> [1]
|  |  |  |  |  |  |  |     |  |     |  └> []
|  |  |  |  |  |  |  |     |  |     └> [2]
|  |  |  |  |  |  |  |     |  |        └> []
|  |  |  |  |  |  |  |     |  └> [22]
|  |  |  |  |  |  |  |     |     ├> []
|  |  |  |  |  |  |  |     |     ├> [12]
|  |  |  |  |  |  |  |     |     |  ├> []
|  |  |  |  |  |  |  |     |     |  ├> [1]
|  |  |  |  |  |  |  |     |     |  |  └> []
|  |  |  |  |  |  |  |     |     |  └> …
|  |  |  |  |  |  |  |     |     └> …
|  |  |  |  |  |  |  |     └> …
|  |  |  |  |  |  |  └> …
|  |  |  |  |  |  └> …
|  |  |  |  |  └> …
|  |  |  |  └> …
|  |  |  └> …
|  |  └> …
|  └> …
└> …"
`;

exports[`array (integration) should be able to shrink [2, 4, 8] given constraints minLength:2 1`] = `
"[2,4,8]
├> [4,8]
|  ├> [2,8]
|  |  ├> [0,8]
|  |  |  ├> [0,6]
|  |  |  |  ├> [0,3]
|  |  |  |  |  └> [0,0]
|  |  |  |  └> [0,4]
|  |  |  |     ├> [0,0]
|  |  |  |     └> [0,1]
|  |  |  └> [0,7]
|  |  |     ├> [0,4]
|  |  |     |  ├> [0,0]
|  |  |     |  └> [0,1]
|  |  |     └> [0,5]
|  |  |        ├> [0,1]
|  |  |        └> [0,2]
|  |  ├> [2,6]
|  |  |  ├> [2,3]
|  |  |  |  └> [2,0]
|  |  |  └> [2,4]
|  |  |     ├> [2,0]
|  |  |     └> [2,1]
|  |  └> [2,7]
|  |     ├> [2,4]
|  |     |  ├> [2,0]
|  |     |  └> [2,1]
|  |     └> [2,5]
|  |        ├> [2,1]
|  |        └> [2,2]
|  ├> [3,8]
|  |  ├> [0,8]
|  |  |  ├> [0,6]
|  |  |  |  ├> [0,3]
|  |  |  |  |  └> [0,0]
|  |  |  |  └> [0,4]
|  |  |  |     ├> [0,0]
|  |  |  |     └> [0,1]
|  |  |  └> [0,7]
|  |  |     ├> [0,4]
|  |  |     |  ├> [0,0]
|  |  |     |  └> [0,1]
|  |  |     └> [0,5]
|  |  |        ├> [0,1]
|  |  |        └> [0,2]
|  |  ├> [1,8]
|  |  |  ├> [1,6]
|  |  |  |  ├> [1,3]
|  |  |  |  |  └> [1,0]
|  |  |  |  └> [1,4]
|  |  |  |     ├> [1,0]
|  |  |  |     └> [1,1]
|  |  |  └> [1,7]
|  |  |     ├> [1,4]
|  |  |     |  ├> [1,0]
|  |  |     |  └> [1,1]
|  |  |     └> [1,5]
|  |  |        ├> [1,1]
|  |  |        └> [1,2]
|  |  ├> [3,6]
|  |  |  ├> [3,3]
|  |  |  |  └> [3,0]
|  |  |  └> [3,4]
|  |  |     ├> [3,0]
|  |  |     └> [3,1]
|  |  └> [3,7]
|  |     ├> [3,4]
|  |     |  ├> [3,0]
|  |     |  └> [3,1]
|  |     └> [3,5]
|  |        ├> [3,1]
|  |        └> [3,2]
|  ├> [4,6]
|  |  ├> [4,3]
|  |  |  └> [4,0]
|  |  └> [4,4]
|  |     ├> [4,0]
|  |     └> [4,1]
|  └> [4,7]
|     ├> [4,4]
|     |  ├> [4,0]
|     |  └> [4,1]
|     └> [4,5]
|        ├> [4,1]
|        └> [4,2]
├> [0,4,8]
|  ├> [4,8]
|  |  ├> [2,8]
|  |  |  ├> [0,8]
|  |  |  |  ├> [0,6]
|  |  |  |  |  ├> [0,3]
|  |  |  |  |  |  └> [0,0]
|  |  |  |  |  └> [0,4]
|  |  |  |  |     ├> [0,0]
|  |  |  |  |     └> [0,1]
|  |  |  |  └> [0,7]
|  |  |  |     ├> [0,4]
|  |  |  |     |  ├> [0,0]
|  |  |  |     |  └> [0,1]
|  |  |  |     └> [0,5]
|  |  |  |        └> …
|  |  |  └> …
|  |  └> …
|  └> …
└> …"
`;

exports[`array (integration) should be able to shrink [2, 4, 8] given constraints minLength:3 1`] = `
"[2,4,8]
├> [0,4,8]
|  ├> [0,2,8]
|  |  ├> [0,0,8]
|  |  |  ├> [0,0,6]
|  |  |  |  ├> [0,0,3]
|  |  |  |  |  └> [0,0,0]
|  |  |  |  └> [0,0,4]
|  |  |  |     ├> [0,0,0]
|  |  |  |     └> [0,0,1]
|  |  |  └> [0,0,7]
|  |  |     ├> [0,0,4]
|  |  |     |  ├> [0,0,0]
|  |  |     |  └> [0,0,1]
|  |  |     └> [0,0,5]
|  |  |        ├> [0,0,1]
|  |  |        └> [0,0,2]
|  |  ├> [0,2,6]
|  |  |  ├> [0,2,3]
|  |  |  |  └> [0,2,0]
|  |  |  └> [0,2,4]
|  |  |     ├> [0,2,0]
|  |  |     └> [0,2,1]
|  |  └> [0,2,7]
|  |     ├> [0,2,4]
|  |     |  ├> [0,2,0]
|  |     |  └> [0,2,1]
|  |     └> [0,2,5]
|  |        ├> [0,2,1]
|  |        └> [0,2,2]
|  ├> [0,3,8]
|  |  ├> [0,0,8]
|  |  |  ├> [0,0,6]
|  |  |  |  ├> [0,0,3]
|  |  |  |  |  └> [0,0,0]
|  |  |  |  └> [0,0,4]
|  |  |  |     ├> [0,0,0]
|  |  |  |     └> [0,0,1]
|  |  |  └> [0,0,7]
|  |  |     ├> [0,0,4]
|  |  |     |  ├> [0,0,0]
|  |  |     |  └> [0,0,1]
|  |  |     └> [0,0,5]
|  |  |        ├> [0,0,1]
|  |  |        └> [0,0,2]
|  |  ├> [0,1,8]
|  |  |  ├> [0,1,6]
|  |  |  |  ├> [0,1,3]
|  |  |  |  |  └> [0,1,0]
|  |  |  |  └> [0,1,4]
|  |  |  |     ├> [0,1,0]
|  |  |  |     └> [0,1,1]
|  |  |  └> [0,1,7]
|  |  |     ├> [0,1,4]
|  |  |     |  ├> [0,1,0]
|  |  |     |  └> [0,1,1]
|  |  |     └> [0,1,5]
|  |  |        ├> [0,1,1]
|  |  |        └> [0,1,2]
|  |  ├> [0,3,6]
|  |  |  ├> [0,3,3]
|  |  |  |  └> [0,3,0]
|  |  |  └> [0,3,4]
|  |  |     ├> [0,3,0]
|  |  |     └> [0,3,1]
|  |  └> [0,3,7]
|  |     ├> [0,3,4]
|  |     |  ├> [0,3,0]
|  |     |  └> [0,3,1]
|  |     └> [0,3,5]
|  |        ├> [0,3,1]
|  |        └> [0,3,2]
|  ├> [0,4,6]
|  |  ├> [0,4,3]
|  |  |  └> [0,4,0]
|  |  └> [0,4,4]
|  |     ├> [0,4,0]
|  |     └> [0,4,1]
|  └> [0,4,7]
|     ├> [0,4,4]
|     |  ├> [0,4,0]
|     |  └> [0,4,1]
|     └> [0,4,5]
|        ├> [0,4,1]
|        └> [0,4,2]
├> [1,4,8]
|  ├> [1,2,8]
|  |  ├> [1,0,8]
|  |  |  ├> [1,0,6]
|  |  |  |  ├> [1,0,3]
|  |  |  |  |  └> [1,0,0]
|  |  |  |  └> [1,0,4]
|  |  |  |     ├> [1,0,0]
|  |  |  |     └> [1,0,1]
|  |  |  └> [1,0,7]
|  |  |     ├> [1,0,4]
|  |  |     |  ├> [1,0,0]
|  |  |     |  └> [1,0,1]
|  |  |     └> [1,0,5]
|  |  |        ├> [1,0,1]
|  |  |        └> …
|  |  └> …
|  └> …
└> …"
`;

exports[`array (integration) should be able to shrink [2, 4, 8] given constraints minLength:undefined 1`] = `
"[2,4,8]
├> []
├> [4,8]
|  ├> [8]
|  |  ├> []
|  |  ├> [6]
|  |  |  ├> []
|  |  |  ├> [3]
|  |  |  |  ├> []
|  |  |  |  └> [0]
|  |  |  |     └> []
|  |  |  └> [4]
|  |  |     ├> []
|  |  |     ├> [0]
|  |  |     |  └> []
|  |  |     └> [1]
|  |  |        └> []
|  |  └> [7]
|  |     ├> []
|  |     ├> [4]
|  |     |  ├> []
|  |     |  ├> [0]
|  |     |  |  └> []
|  |     |  └> [1]
|  |     |     └> []
|  |     └> [5]
|  |        ├> []
|  |        ├> [1]
|  |        |  └> []
|  |        └> [2]
|  |           └> []
|  ├> [2,8]
|  |  ├> [8]
|  |  |  ├> []
|  |  |  ├> [6]
|  |  |  |  ├> []
|  |  |  |  ├> [3]
|  |  |  |  |  ├> []
|  |  |  |  |  └> [0]
|  |  |  |  |     └> []
|  |  |  |  └> [4]
|  |  |  |     ├> []
|  |  |  |     ├> [0]
|  |  |  |     |  └> []
|  |  |  |     └> [1]
|  |  |  |        └> []
|  |  |  └> [7]
|  |  |     ├> []
|  |  |     ├> [4]
|  |  |     |  ├> []
|  |  |     |  ├> [0]
|  |  |     |  |  └> []
|  |  |     |  └> [1]
|  |  |     |     └> []
|  |  |     └> [5]
|  |  |        ├> []
|  |  |        ├> [1]
|  |  |        |  └> []
|  |  |        └> [2]
|  |  |           └> []
|  |  ├> [0,8]
|  |  |  ├> [8]
|  |  |  |  ├> []
|  |  |  |  ├> [6]
|  |  |  |  |  ├> []
|  |  |  |  |  ├> [3]
|  |  |  |  |  |  ├> []
|  |  |  |  |  |  └> [0]
|  |  |  |  |  |     └> []
|  |  |  |  |  └> [4]
|  |  |  |  |     ├> []
|  |  |  |  |     ├> [0]
|  |  |  |  |     |  └> []
|  |  |  |  |     └> [1]
|  |  |  |  |        └> []
|  |  |  |  └> [7]
|  |  |  |     ├> []
|  |  |  |     ├> [4]
|  |  |  |     |  ├> []
|  |  |  |     |  ├> [0]
|  |  |  |     |  |  └> []
|  |  |  |     |  └> [1]
|  |  |  |     |     └> []
|  |  |  |     └> [5]
|  |  |  |        ├> []
|  |  |  |        ├> [1]
|  |  |  |        |  └> []
|  |  |  |        └> [2]
|  |  |  |           └> []
|  |  |  ├> [0]
|  |  |  |  └> []
|  |  |  ├> [0,6]
|  |  |  |  ├> [6]
|  |  |  |  |  ├> []
|  |  |  |  |  ├> [3]
|  |  |  |  |  |  ├> []
|  |  |  |  |  |  └> [0]
|  |  |  |  |  |     └> []
|  |  |  |  |  └> [4]
|  |  |  |  |     ├> []
|  |  |  |  |     └> …
|  |  |  |  └> …
|  |  |  └> …
|  |  └> …
|  └> …
└> …"
`;
